### PR TITLE
feat: shrink package size, improve dev deps, esm and cjs support

### PR DESCRIPTION
**Pull request by jacoobes**
This merge 
- esm and cjs support
- better development experience 
     - uses tsup for faster build times
     - scripts directory
     - more automation
- shrink install size ( -8 Mbs ) !
- improves comments

### CODE REVIEW

**Comment by jacoobes** [comment]
This will be the final branch until version 1.0.0

File: package.json [hunk 1]
```diff
@@ -21,17 +31,17 @@
   "author": "SernDevs",
   "license": "MIT",
   "dependencies": {
-    "discord.js": "^14.0.1",
     "rxjs": "^7.5.6",
     "ts-pattern": "^4.0.2",
-    "ts-results": "^3.3.0"
+    "ts-results-es": "^3.5.0"
```
Comment (EvolutionX-10) [inline package.json:36]: poggers

File: src/handler/structures/errors.ts [hunk 2]
```diff
@@ -1,9 +1,34 @@
+/**
+ * @enum { string }
+ */
 export enum SernError {
+    /**
+     * Throws when registering an invalid module.
+     * This means it is undefined or an invalid command type was provided
+     */
     NonValidModuleType = 'Detected an unknown module type',
```
Comment (EvolutionX-10) [inline src/handler/structures/errors.ts:9]: `NonValidModuleType` or `InvalidModuleType`
what do you think

**Review by EvolutionX-10 — APPROVED** [review]
GG you did it! cheers
Have a look at comment

File: src/handler/structures/errors.ts (hunk 2 above)
Comment (jacoobes) [reply src/handler/structures/errors.ts:9]: InvalidModuleType is better

Comment (jacoobes) [reply src/handler/structures/errors.ts:9]: This would be a breaking change, correct?

Comment (EvolutionX-10) [reply src/handler/structures/errors.ts:9]: does it break code of current cjs users?

Comment (EvolutionX-10) [reply src/handler/structures/errors.ts:9]: i assume you'll rename and commit, resolving this @jacoobes

File: package.json [hunk 3]
```diff
@@ -1,13 +1,23 @@
...
+    ".": {
+      "import": "./dist/esm/index.mjs",
+      "require": "./dist/cjs/index.cjs"
+    }
+  },
   "scripts": {
-    "compile": "tsc",
-    "watch": "tsc -w",
+    "watch": "tsup --watch --dts",
+    "clean-modules": "rimraf node_modules/ && npm install",
```
Comment (Murtatrxx) [inline package.json:16]: What's rimraf though? 🤔

File: scripts/mkjson.mjs [hunk 4]
```diff
@@ -0,0 +1,15 @@
+import { writeFile } from 'fs/promises'
+import { join } from 'path';
+//A quick script to regenerate package.jsons for each cjs and esm after tsup cleans distributions
```
Comment (Murtatrxx) [inline scripts/mkjson.mjs:3]: For future, please add a space between // and comment, it just looks nicer.

File: src/handler/events/observableHandling.ts [hunk 5]
```diff
@@ -2,31 +2,11 @@ import type { Message } from 'discord.js';
...
-                        subscriber.next(mod!);
-                    } else {
-                        return throwError(() => SernError.MismatchModule);
-                    }
-                },
-                error: e => subscriber.error(e),
-                complete: () => subscriber.complete(),
-            });
-        });
-}
```
Comment (Murtatrxx) [inline src/handler/events/observableHandling.ts:29]: this isn't needed anymore??

File: src/handler/plugins/plugin.ts [hunk 6]
```diff
@@ -7,14 +7,14 @@
  * The goal of plugins is to organize commands and
  * provide extensions to repetitive patterns
  * examples include refreshing modules,
- * categorizing commands, cooldowns, permissions, etc.
```
Comment (Murtatrxx) [inline src/handler/plugins/plugin.ts:10]: This really does not matter idk + idc

File: src/handler/sern.ts [hunk 7]
```diff
@@ -1,5 +1,5 @@
 import type Wrapper from './structures/wrapper';
-import { Err, Ok } from 'ts-results';
+import { Err, Ok } from 'ts-results-es';
```
Comment (Murtatrxx) [inline src/handler/sern.ts:2]: For esm comparability yeah 😎😎

File: src/handler/structures/enums.ts [hunk 8]
```diff
@@ -1,30 +1,119 @@
 /**
```
Comment (Murtatrxx) [inline src/handler/structures/enums.ts:1]: Good documentation though

File: src/handler/structures/wrapper.ts [hunk 9]
```diff
@@ -3,13 +3,8 @@ import type SernEmitter from '../sernEmitter';
 import type { EventModule } from './module';
 
 /**
- * An object to be passed into Sern.Handler constructor.
+ * An object to be passed into init function.
```
Comment (Murtatrxx) [inline src/handler/structures/wrapper.ts:6]: consider using Sern#init()

**Review by Murtatrxx — APPROVED** [review]
Looks good at all, needs small improvements.

File: scripts/mkjson.mjs (hunk 4 above)
Comment (jacoobes) [reply scripts/mkjson.mjs:3]: 👍

File: package.json (hunk 3 above)
Comment (jacoobes) [reply package.json:16]: rm rf but for windows xD

File: src/handler/structures/errors.ts (hunk 2 above)
Comment (jacoobes) [reply src/handler/structures/errors.ts:9]: yes

File: .npmignore [hunk 10]
```diff
@@ -1,5 +1,4 @@
 src/
-tsconfig.json
 docs/
```
Comment (Murtatrxx) [inline .npmignore:2]: We don't need that anymore

**Review by Murtatrxx — CHANGES_REQUESTED** [review]
Only one thing left, and then it's good to go so far 😃

File: .npmignore (hunk 10 above)
Comment (jacoobes) [reply .npmignore:2]: removed

**Review by Murtatrxx — APPROVED** [review]

**Comment by jacoobes** [comment]
cheers

**Pull request merged**